src/main/java/seedu/todo/logic/commands/TagCommand.java

```diff
@@ -0,0 +1,238 @@
+package seedu.todo.logic.commands;
+
+import com.google.common.collect.ImmutableList;
+import org.ocpsoft.prettytime.shade.org.apache.commons.lang.BooleanUtils;
+import seedu.todo.commons.exceptions.IllegalValueException;
+import seedu.todo.commons.exceptions.ValidationException;
+import seedu.todo.commons.util.CollectionUtil;
+import seedu.todo.commons.util.StringUtil;
+import seedu.todo.logic.arguments.Argument;
+import seedu.todo.logic.arguments.IntArgument;
+import seedu.todo.logic.arguments.Parameter;
+import seedu.todo.logic.arguments.StringArgument;
+import seedu.todo.model.ErrorBag;
+import seedu.todo.model.tag.Tag;
+import seedu.todo.model.task.ImmutableTask;
+
+import java.util.stream.Collectors;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Set;
+import java.util.Arrays;
+
+//@@author A0135805H
+/**
+ * This class handles all tagging command
+ */
+public class TagCommand extends BaseCommand {
+    /* Constants */
+    private static final String VERB = "tagged";
+    private static final String TAG_VALIDATION_REGEX = "([A-Za-z0-9_-])+";
```

Comment:
I don't think this regex would work - you didn't anchor the regex, so it might succeed for something like `a@` - there is at least one word character, so the regex succeed. The capturing group is also unnecessary. Try `^[\\w\\d_-]+$` instead. Might also be a good idea to compile this.